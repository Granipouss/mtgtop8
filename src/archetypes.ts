import { getPage } from './utils';

export type ShortDeck = {
  id: number;
  eventId: number;
  name: string;
};

export type Archetype = { id: number; name: string; decks: ShortDeck[] };

const parseName = ($el: Cheerio): string =>
  $el
    .eq(1)
    .text()
    .trim()
    .replace(/ decks$/, '');

const parseDeck = ($el: Cheerio): ShortDeck => {
  const regexp = /^<a href="event\?e=(\d+)&amp;d=(\d+)&amp;f=MO">(.+)<\/a>$/;
  const match = $el.find('td').eq(1).html()?.match(regexp);
  if (!match) throw new Error(`Cannot parse deck: ${$el.text()}`);

  const eventId = parseInt(match[1], 10);
  const id = parseInt(match[2], 10);
  const name = match[3].trim();

  return { eventId, id, name };
};

export const getArchetype = async (
  id: number,
  page = 1
): Promise<Archetype> => {
  const $ = await getPage(
    `https://www.mtgtop8.com/archetype?a=${id}&current_page=${page}`
  );

  const archetype: Archetype = {
    id,
    name: parseName($('.w_title')),
    decks: [],
  };

  $('table table[align=center] .hover_tr').each((_, el) => {
    archetype.decks.push(parseDeck($(el)));
  });

  return archetype;
};
