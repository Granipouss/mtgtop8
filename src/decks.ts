import { getPage } from './utils';

export type ShortArchetype = { name: string; id: number };

export type CardGroup = 'mainboard' | 'sideboard' | 'commander';

export type Card = { count: number; name: string; isCompanion?: boolean };

export type Deck = {
  id: number;
  eventId: number;
  format: string;
  name: string;
  archetype: ShortArchetype;
  player: string;
  result: string;
} & {
  [group in CardGroup]: Card[];
};

const parseFormat = ($el: Cheerio): string => {
  const match = $el.html()?.match(/(.+) <div/);
  if (!match) throw new Error(`Cannot parse format: ${$el.text()}`);
  return match[1].trim();
};

const parseArchetype = ($el: Cheerio): ShortArchetype => {
  const regexp = /^<a href="archetype\?a=(\d+)">(.+)<br>decks<\/a>$/;
  const match = $el.html()?.match(regexp);
  if (!match) throw new Error(`Cannot parse archetype: ${$el.text()}`);

  const id = parseInt(match[1], 10);
  const name = match[2].trim();

  return { id, name };
};

const parseCard = ($el: Cheerio): Card => {
  const text = $el.text();
  const match = text.match(/^(\d+) (.+)$/);
  if (!match) throw new Error(`Cannot parse card: ${text}`);

  const count = parseInt(match[1], 10);
  const name = match[2].trim();
  const isCompanion = $el.find('[id="companion_%%%"]').length > 0;

  const card: Card = { count, name };
  if (isCompanion) card.isCompanion = true;

  return card;
};

const parseGroup = ($el: Cheerio): CardGroup => {
  switch ($el.text()) {
    case 'COMMANDER':
      return 'commander';
    case 'SIDEBOARD':
      return 'sideboard';
    default:
      return 'mainboard';
  }
};

export const getDeck = async (eventId: number, id: number): Promise<Deck> => {
  const $ = await getPage(`https://www.mtgtop8.com/event?e=${eventId}&d=${id}`);

  if ($.html().includes('No event could be found.')) {
    throw new Error(`No deck ${id} for event ${eventId}`);
  }

  const deck: Deck = {
    id,
    eventId,
    name: $('table .chosen_tr :not([align=right]) .topic').text().trim(),
    player: $('table .chosen_tr [align=right] .topic').text().trim(),
    result: $('table .chosen_tr [align=center]').text().trim(),
    format: parseFormat($('table table td[align="center"]').first()),
    archetype: parseArchetype($('.Nav_link').eq(1)),
    mainboard: [],
    sideboard: [],
    commander: [],
  };

  let currentGroup: CardGroup = 'mainboard';
  $('table table table td td').each((_, el) => {
    const isCard = $('.hover_tr', el).length > 0;
    if (isCard) {
      deck[currentGroup].push(parseCard($(el)));
    } else {
      currentGroup = parseGroup($(el));
    }
  });

  return deck;
};
