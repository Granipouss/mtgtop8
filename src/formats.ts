import { getPage } from './utils';

export type Format = { name: string; code: string };

export const getFormats = async (): Promise<Format[]> => {
  const $ = await getPage(`https://www.mtgtop8.com/`);

  const formats: Format[] = [];

  $('table')
    .eq(1)
    .find('td')
    .each((_, el) => {
      const link = $(el).find('a').attr('href');
      const linkRegexp = /https:\/\/www\.mtgtop8\.com\/format\?f=(.+)/;
      const linkMatch = link?.match(linkRegexp);
      if (!linkMatch) return;
      const code = linkMatch[1].trim().toUpperCase();

      const alt = $(el).find('img').attr('id');
      if (!alt) return;
      const name = alt
        .replace(/^menu_/, '')
        .replace(/_/g, ' ')
        .toLowerCase();

      formats.push({ name, code });
    });

  $('#other_tooltip_content div').each((_, el) => {
    const link = $(el).find('a').attr('href');
    const linkRegexp = /https:\/\/www\.mtgtop8\.com\/format\?f=(.+)/;
    const linkMatch = link?.match(linkRegexp);
    if (!linkMatch) return;
    const code = linkMatch[1].trim().toUpperCase();

    const alt = $(el).find('img').attr('id');
    if (!alt) return;
    const name = alt
      .replace(/^menu_/, '')
      .replace(/_/g, ' ')
      .toLowerCase();

    formats.push({ name, code });
  });

  return formats;
};
