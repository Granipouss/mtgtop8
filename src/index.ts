export * from './format';
export * from './archetype';
export * from './event';
export * from './deck';
export * from './search';
