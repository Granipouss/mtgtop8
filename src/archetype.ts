import { getPage, url, match } from './utils';

import { ShortDeck } from './deck';

export type ShortArchetype = { id: number; name: string };

export class Archetype implements ShortArchetype {
  public static async get(id: number, page = 1): Promise<Archetype> {
    const $ = await getPage(url.archetype(id, page));
    return new Archetype($);
  }

  id: number;
  name: string;

  decks: ShortDeck[];

  private constructor($: CheerioStatic) {
    this.id = parseInt(match($('table table option').attr('value'), /archetype\?a=(\d+)&f=.+&meta=.+/)[0], 10);

    this.name = $('table table').eq(1).find('tr:first-child').text().replace('decks', '').trim();

    this.decks = $('table table[align=center] .hover_tr')
      .get()
      .map((el) => {
        const [eventString, deckString, name] = match(
          $(el).find('td').eq(1).html() || '',
          /^<a href="event\?e=(\d+)&amp;d=(\d+)&amp;f=MO">(.+)<\/a>$/
        );

        return {
          eventId: parseInt(eventString, 10),
          id: parseInt(deckString, 10),
          name: name.trim(),
        };
      });
  }
}
