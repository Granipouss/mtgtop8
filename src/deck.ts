import { getPage, url, match, ListOf, WithCount, parseDeckLink } from './utils';

import { ShortArchetype } from './archetype';
import { ShortFormat } from './format';

export type ShortDeck = { eventId: number; id: number; name: string };

export type CardGroup = 'mainboard' | 'sideboard' | 'commander';

export type Card = { name: string; isCompanion?: boolean };

export class Deck implements ShortDeck {
  public static async get(eventId: number, id: number): Promise<Deck> {
    const $ = await getPage(url.deck(eventId, id));

    // FIXME: check for deck also
    if ($.html().includes('No event could be found.')) {
      throw new Error(`No deck ${id} for event ${eventId}`);
    }

    return new Deck($);
  }

  eventId: number;
  id: number;
  name: string;

  format: ShortFormat;
  archetype: ShortArchetype;
  player: string;
  result: string;

  mainboard: ListOf<Card> = [];
  sideboard: ListOf<Card> = [];
  commander: ListOf<Card> = [];

  private constructor($: CheerioStatic) {
    const chosenLinkData = parseDeckLink($('.chosen_tr a').attr('href'));

    this.eventId = chosenLinkData.event;

    this.id = chosenLinkData.deck;

    this.name = $('table .chosen_tr :not([align=right]) .topic').text().trim();

    this.format = {
      name: match($('table table td[align="center"]').html() || '', /(.+) <div/)[0].trim(),
      code: chosenLinkData.format,
    };

    const [archetypeId, archetypeName] = match(
      $('.Nav_link').eq(1).html() || '',
      /^<a href="archetype\?a=(\d+)">(.+)<br>decks<\/a>$/
    );

    this.archetype = {
      id: parseInt(archetypeId, 10),
      name: archetypeName.trim(),
    };

    this.player = $('table .chosen_tr [align=right] .topic').text().trim();

    this.result = $('table .chosen_tr [align=center]').text().trim();

    let currentGroup: CardGroup = 'mainboard';
    $('table table table td td').each((_, el) => {
      const isCard = $('.hover_tr', el).length > 0;
      if (isCard) {
        this[currentGroup].push(parseCard($(el)));
      } else {
        currentGroup = parseGroup($(el));
      }
    });
  }
}

const parseCard = ($el: Cheerio): WithCount<Card> => {
  const text = $el.text();
  const match = text.match(/^(\d+) (.+)$/);
  if (!match) throw new Error(`Cannot parse card: ${text}`);

  const count = parseInt(match[1], 10);
  const name = match[2].trim();
  const isCompanion = $el.find('[id="companion_%%%"]').length > 0;

  const card: WithCount<Card> = { count, name };
  if (isCompanion) card.isCompanion = true;

  return card;
};

const parseGroup = ($el: Cheerio): CardGroup => {
  switch ($el.text()) {
    case 'COMMANDER':
      return 'commander';
    case 'SIDEBOARD':
      return 'sideboard';
    default:
      return 'mainboard';
  }
};
