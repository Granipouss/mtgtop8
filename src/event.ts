import { getPage, url, match, parseDeckLink, isDefined } from './utils';
import { ShortFormat } from './format';
import { ShortDeck } from './deck';

export type ShortEvent = { id: number; name: string };

type WithRanks<T> = T & { ranks: number[] };

export class Event implements ShortEvent {
  public static async get(id: number): Promise<Event> {
    const $ = await getPage(url.event(id));

    if ($.html().includes('No event could be found.')) {
      throw new Error(`No event ${id}`);
    }

    return new Event($);
  }

  id: number;
  name: string;

  format: ShortFormat;
  rank: number;
  playerCount: number;
  date: Date;
  decks: WithRanks<ShortDeck>[] = [];

  private constructor($: CheerioStatic) {
    const chosenLinkData = parseDeckLink($('.chosen_tr a').attr('href'));

    this.id = chosenLinkData.event;

    this.name = $('.w_title td:first-child').text().trim();

    this.format = {
      name: match($('table table td[align="center"]').html() || '', /(.+) <div/)[0].trim(),
      code: chosenLinkData.format,
    };

    const stars = $('img[src="/graph/star.png"]').length;
    const bigstars = $('img[src="/graph/bigstar.png"]').length;
    this.rank = 10 * bigstars + stars;

    const [playerCount, date] = match($('table table td[align="center"]').html() || '', /(\d+) players - (.+)/);

    this.playerCount = parseInt(playerCount, 10);

    this.date = new Date(`${date}Z`);

    this.decks = $('.hover_tr, .chosen_tr')
      .get()
      .map((el) => {
        try {
          const linkData = parseDeckLink($(el).find('a').first().attr('href'));
          return {
            ranks: $(el).find('div:first-child').text().split('-').map(Number),
            eventId: linkData.event,
            id: linkData.deck,
            name: $(el).find('a').first().text(),
          };
        } catch (error) {
          return;
        }
      })
      .filter(isDefined);
  }
}
