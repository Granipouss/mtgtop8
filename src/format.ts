import { getPage, match, ListOf, isDefined, url } from './utils';

import { ShortArchetype } from './archetype';
import { ShortEvent } from './event';

export type ShortFormat = { name: string; code: string };

export class Format implements ShortFormat {
  public static async get(code: string): Promise<Format> {
    const $ = await getPage(url.format(code));

    if (getTableRows($, /Metagame Breakdown/).length === 1) {
      throw new Error(`No format for code ${code}`);
    }

    return new Format($);
  }

  public static async getAll(): Promise<ShortFormat[]> {
    const $ = await getPage(url.base());
    return [...$('table').eq(1).find('td').get(), ...$('#other_tooltip_content div').get()]
      .map((el) => {
        try {
          const [code] = match($(el).find('a').attr('href'), /https:\/\/www\.mtgtop8\.com\/format\?f=(.+)/);

          const alt = $(el).find('img').attr('id');
          if (!alt) return;
          const name = alt
            .replace(/^menu_/, '')
            .replace(/_/g, ' ')
            .toLowerCase();

          return { name, code };
        } catch (error) {
          return;
        }
      })
      .filter(isDefined);
  }

  public name: string;
  public code: string;

  public archetypes: ListOf<ShortArchetype> = [];

  private constructor($: CheerioStatic) {
    this.code = match($('table table .hover_tr a').attr('href'), /archetype\?a=.+&meta=.+&f=(.+)/)[0];

    this.name = $('table table a[target=_blank]')
      .text()
      .replace(/ Construction Rules$/, '')
      .toLowerCase()
      .trim();

    this.archetypes = getTableRows($, /Metagame Breakdown/)
      .filter((el) => $(el).is('.hover_tr'))
      .map((el) => {
        const [idString] = match($(el).find('a').attr('href'), /archetype\?a=(\d+)&meta=.+&f=.+/);
        return {
          id: parseInt(idString, 10),
          name: $(el).find('a').text().trim(),
          count: parseInt($(el).find('[align=right]').first().text(), 10),
        };
      });
  }

  async getEvents(page = 1): Promise<ShortEvent[]> {
    const $ = await getPage(url.format(this.code, page));
    return getTableRows($, /^(Events \d+ to \d+|Last \d+ events)$/).map((el) => {
      const [idString] = match($(el).find('a').attr('href'), /event\?e=(\d+)&f=.+/);
      return {
        id: parseInt(idString, 10),
        name: $(el).find('a').text().trim(),
      };
    });
  }
}

const getTableRows = ($: CheerioStatic, title: RegExp): CheerioElement[] => {
  try {
    const table: CheerioElement | undefined = $('table table')
      .get()
      .find((el) => $('tr:first-child', el).text().match(title));
    return table ? $(table).find('tr').get().splice(1) : [];
  } catch (error) {
    return [];
  }
};
