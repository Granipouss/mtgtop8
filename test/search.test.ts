import { Search, Level } from '../src';

describe('Search', () => {
  it('should search for nice decks', async () => {
    const response = await Search.make()
      .setDeckName('faerie')
      .setFormat('MO')
      .setLevels([Level.Professional, Level.Competitive])
      .setDateEnd(new Date('2018-04-16'))
      .run();

    expect(response).toMatchSnapshot();
  });

  it('should not fail', async () => {
    const response = await Search.make().setDateStart(new Date('1990-04-16')).setDateEnd(new Date('1990-04-16')).run();

    expect(response).toEqual({ count: 0, results: [] });
  });
});
