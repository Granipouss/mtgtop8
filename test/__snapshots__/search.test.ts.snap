// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should search for nice decks 1`] = `
Object {
  "count": 43,
  "results": Array [
    Object {
      "eventId": 16771,
      "id": 303732,
      "name": "UB Faeries",
    },
    Object {
      "eventId": 16709,
      "id": 303318,
      "name": "Ub Faeries",
    },
    Object {
      "eventId": 16542,
      "id": 302304,
      "name": "Faeries",
    },
    Object {
      "eventId": 16084,
      "id": 299101,
      "name": "Faeries",
    },
    Object {
      "eventId": 16429,
      "id": 301513,
      "name": "Ub Faeries",
    },
    Object {
      "eventId": 14765,
      "id": 288904,
      "name": "Faeries",
    },
    Object {
      "eventId": 14511,
      "id": 287018,
      "name": "Faeries",
    },
    Object {
      "eventId": 13636,
      "id": 280254,
      "name": "Faeries",
    },
    Object {
      "eventId": 11328,
      "id": 264321,
      "name": "Faeries",
    },
    Object {
      "eventId": 10980,
      "id": 262515,
      "name": "Faeries",
    },
    Object {
      "eventId": 9545,
      "id": 254811,
      "name": "Faeries",
    },
    Object {
      "eventId": 9545,
      "id": 254813,
      "name": "Faeries",
    },
    Object {
      "eventId": 8087,
      "id": 246592,
      "name": "Faeries",
    },
    Object {
      "eventId": 7999,
      "id": 246035,
      "name": "Faeries",
    },
    Object {
      "eventId": 7862,
      "id": 245274,
      "name": "Faeries",
    },
    Object {
      "eventId": 7858,
      "id": 245251,
      "name": "Faeries",
    },
    Object {
      "eventId": 7727,
      "id": 244465,
      "name": "Faeries",
    },
    Object {
      "eventId": 6980,
      "id": 240112,
      "name": "Faeries",
    },
    Object {
      "eventId": 2776,
      "id": 217953,
      "name": "Mono U Faeries",
    },
    Object {
      "eventId": 2611,
      "id": 216970,
      "name": "Faeries",
    },
    Object {
      "eventId": 2622,
      "id": 217042,
      "name": "Faeries",
    },
    Object {
      "eventId": 2597,
      "id": 216887,
      "name": "UR Faeries",
    },
    Object {
      "eventId": 2594,
      "id": 216860,
      "name": "Faeries",
    },
    Object {
      "eventId": 2600,
      "id": 216909,
      "name": "UR Faeries",
    },
    Object {
      "eventId": 2572,
      "id": 216712,
      "name": "Faeries",
    },
  ],
}
`;
